cleanup the style guide

yjbanov opened the pull request:
- do not repeat the Dart style guide
- use DO/DON'T/AVOID/PREVER headings similar to Effective Dart's
- consistent code example formatting
- consistent GOOD/BAD example formatting
- consistent heading syntax
- consistent line length
- more code samples

eseidelGoogle commented:
Once this lands, we should also consider moving it to flutter/flutter or flutter/website.  It's really strangely placed atm.

Hixie commented on sky/specs/style-guide.md:
```diff
@@ -1,5 +1,4 @@
-Flutter Style Guide
-===================
+# Flutter Style Guide
```
Prefer === because it makes it more readable in text mode.

yjbanov replied:
Me too, but I don't know if this style supports anything beyond 2nd-level headings, so I decided to switch everything to `#` for consistency sake.

Hixie replied:
PREFER readability over consistency. :-)

yjbanov replied:
Consistency contributes to readability big time. But we're talking about a couple headlines here, so I'll revert the change.

Hixie commented on sky/specs/style-guide.md:
```diff
@@ -8,44 +7,114 @@ that everyone, whether reading the code for the first time or
 secondary goal is avoiding arguments when there are disagreements.
 
-All languages
--------------
```
ditto here

yjbanov replied:
done

yjbanov commented:
We should probably kill the stubs for Java, Objective-C and C++ at the bottom, if we want to move it to flutter/flutter.

Hixie commented on sky/specs/style-guide.md:
```diff
@@ -8,44 +7,114 @@ that everyone, whether reading the code for the first time or
 
-Avoid checking in commented-out code. It will bitrot too fast to be
-useful, and will confuse people maintaining the code.
+### AVOID checking in commented-out code
```
The all-caps "AVOID" and "DON'T" and so on is pretty ugly. Can we do something more typographically pleasing?

In general having headings makes sense (the existing style guide is a wall of text with no headings organising it), but I'm not sure having the rules in the headings make sense. Why, for example, does "AVOID checking in comments that ask questions" have a heading, but "If commenting on a workaround due to a bug, also leave a link to the bug and a TODO to clean it up when the bug is fixed." not? Why is the heading "AVOID checking in comments that ask questions" rather than "DO find the answers to the questions, or describe the confusion, including references to where you found answers"?

Should the headings be more things like "Comments", "Line breaking" etc?

yjbanov replied:
The answer to everything is: be consistent with the standard style guide. This style guide is a delta on top of the standard Dart guide, which we recommend that people read. Therefore, it makes a lot of sense to preserve the same style. Some additional points:

## Why rules in the headings

- they make it easy to quickly scan the guide and familiarizing yourself with the main points
- they make it easy to copy and paste guidelines into a code review comment

## What's with the choice of headlines

Like in the standard guide, each headline attempts to make the most important succinctly. The paragraphs then expand on the point, give examples, and mention asides. We can certainly switch things up as necessary. I by no means insist that my choice of headlines is the most optimal. I'm simply bootstrapping an idea.

Hixie replied:
I think the Dart style guide using capital letters in the headings is pretty ugly, but the way the Dart style guide is organised they first have headings for each topic, and then have headings that barely look like more than emphasised paragraphs.

I think we could find a middle ground where we have topic headings, then subheadings for each major guideline, but where the guidelines are written as full sentences with normal capitalisation (rather than shoehorning them into this "DO that" "PREFER this" style).

Hixie commented on sky/specs/style-guide.md:
````diff
@@ -8,44 +7,114 @@ that everyone, whether reading the code for the first time or
+
+Example:
+
+```dart
````
maybe skip the `dart` here since it just makes the text gray instead of black

yjbanov replied:
done

Hixie commented on sky/specs/style-guide.md:
```diff
@@ -8,44 +7,114 @@ that everyone, whether reading the code for the first time or
+// specification, it should be 3.0. We split the difference and went with
+// 2.5, because we didn't know what else to do.
+
+// TODO: Converting color to RGB because class Color doesn't support hex yet.
```
TODOs should have usernames, as in `TODO(ianh)`

yjbanov replied:
done

Hixie commented on sky/specs/style-guide.md:
````diff
@@ -8,44 +7,114 @@ that everyone, whether reading the code for the first time or
+//       See http://link/to/a/bug/123
+```
+
+## Dart
````
Use --- for level-2 headings

yjbanov replied:
done

Hixie commented on sky/specs/style-guide.md:
```diff
@@ -8,44 +7,114 @@ that everyone, whether reading the code for the first time or
 
-Use assert()s liberally to describe the contracts that you expect your
-code to follow.
+### DON'T use dartfmt (yet)
```
Promoting this "don't" from a parenthetical to a heading seems a bit drastic. We don't really want to draw attention to the problems of using dartfmt, it isn't super polite to the people maintaining dartfmt.

yjbanov replied:
done

Hixie commented on sky/specs/style-guide.md:
```diff
@@ -8,44 +7,114 @@ that everyone, whether reading the code for the first time or
+http://dartbug.com/new.
+
+
+### PREFER using asserts to enforce contracts and invariants
```
prefer over what?

I think I preferred the original phrasing of "use asserts liberally", because it gives you a concrete direction.

yjbanov replied:
done

Hixie commented on sky/specs/style-guide.md:
```diff
@@ -8,44 +7,114 @@ that everyone, whether reading the code for the first time or
-doubles and strings are prefixed with k. Prefer using a local const
-or a static const in a relevant class than using a global constant.
+
+### DO use the Dart Analyzer before submitting code
```
This ("do use...") is grammatically dubious.

yjbanov replied:
Changed to "DO run"

Hixie commented on sky/specs/style-guide.md:
````diff
@@ -8,44 +7,114 @@ that everyone, whether reading the code for the first time or
+```
+
+
+### DO camel-case constant doubles and strings, and begin with prefix "k"
````
This is grammatically dubious too. I think this would end up being more readable if we used topic headings (e.g. "Identifiers") and then described the style in full sentences rather than trying to shoe-horn the style description into the style used by the dart style guide.

yjbanov replied:
Done. What's the value of this guideline, btw?

Hixie replied:
Which part?

yjbanov replied:
The "k" prefix?

Hixie replied:
It's just what we've organically grown to doing. I can give you post-hoc justifications: it makes the constants easier to distinguish from possibly-side-effect-triggering or possibly-varying-over-time getters, it groups all the constants in alphabetic lists, it makes it obvious when reading the code what is just a convenient named constant vs what is a computed value. But mostly it's just an arbitrary style thing. I think the reason I like it most is that classes, variables, constants, and methods each have a clear way to be recognised - UpperCamelCase, lowerCamelCase, kPrefixedCamelCase, and lowerCamelCaseWithParens(). (There is, unfortunately, still no way to distinguish final variables and non-final variables and getters, which can sometimes lead to some confusion when reading the code.)

yjbanov replied:
Why don't we apply it to things other than doubles and strings then? Also, what to do when something that was a constant ceases to be a constant. Break API?

Hixie replied:
We do apply it to other types actually. We should probably change that part of the guide. I'm looking at code right now that uses it for Durations and a Map of ints to BoxShadows.

Changing a constant to something that isn't a constant is a breaking change regardless of what happens to the identifier name.

yjbanov replied:
Fixed

Hixie commented on sky/specs/style-guide.md:
````diff
@@ -8,44 +7,114 @@ that everyone, whether reading the code for the first time or
+```
+
+
+### CONSIDER minimizing the visibility scope of constants
````
you have to do more than consider it. :-)

yjbanov replied:
done

Hixie commented on sky/specs/style-guide.md:
```diff
@@ -255,21 +372,44 @@ this pattern. If for some reason you don't want to use 'value', use
+As per normal Dart conventions, a package should have a single import that
+reexports all of its API.
+
+For example, [rendering.dart](https://github.com/flutter/engine/blob/master/sky/packages/sky/lib/rendering.dart)
```
The original here tries to distinguish "normative" style description from "informative" examples of applying the style. I think we should do that throughout. It makes it clearer what is the actual style guide and what is merely explanatory (and less likely to be rigorously correct).

yjbanov replied:
Reverted. I can see this distinction useful in a spec. For a guide though, I think a more casual style is more appropriate (less styling - less visual noise, shorter more casual text - higher chance that people will actually read it).

Hixie commented:
Overall this is a good improvement but I'm not a fan of the heading style.

yjbanov commented:
Alright, I removed the all-caps DOs and DON'Ts, and added sub-sections. I tried to keep "Do", "Don't", "Prefer", etc where a clear sentence can be formed, and deviated where shoehorning a phrase would be necessary. Having the first word be clear about whether you are encouraged to or discouraged from doing something is useful. I also removed the speculative sections for other languages. It's easy to add them back.

Submitting. We can iterate on it later.

Hixie commented:
LGTM.

(As a general rule please wait until you have an explicit LGTM before checking something in -- in this case, for example, I hadn't finished reviewing it and was waiting for the headings to get resolved before going through it again more thoroughly.)

yjbanov commented:
Sorry, I read your "Overall this is a good improvement" as LGTM. What do you think of adding a label "lgtm" to the engine and flutter repos? I've used it in the past and it's a much more concrete way of LGTM'ing PRs.
